ui/dashboards/src/components/PanelDrawer/panel-editor-model.ts

```diff
@@ -0,0 +1,149 @@
+    return {
+      drawerTitle: 'Add Panel',
+      submitButtonText: 'Add',
+      initialValues: {
+        name: '',
+        description: '',
+        group: groupIndex ?? 0,
+        kind: '',
+        spec: {},
+      },
+      applyChanges: (values) => {
+        const { name, description, group, kind, spec } = values;
+        const panelKey = removeWhiteSpacesAndSpecialCharacters(name);
+        updatePanel(
+          panelKey,
+          {
+            kind: 'Panel',
+            spec: {
+              display: { name, description },
+              plugin: {
+                kind,
+                spec,
+              },
+            },
+          },
+          group
+        );
+      },
+    };
+  }
+
+  // Otherwise we don't have a panel key, so we're trying to edit an existing panel
+  const existingPanel = panels[panelKey];
+
+  // TODO: Can we better express this via the type system on the dashboard store to avoid these states?
+  if (existingPanel === undefined) {
+    throw new Error(`Cannot find existing panel '${panelKey}' to edit`);
+  }
+  if (groupIndex === undefined) {
+    throw new Error(`Cannot edit existing panel '${panelKey}' without its group index`);
+  }
+
+  return {
+    drawerTitle: 'Edit Panel',
+    submitButtonText: 'Apply',
+    initialValues: {
+      name: existingPanel.spec.display.name,
+      description: existingPanel.spec.display.description ?? '',
+      group: groupIndex,
+      kind: existingPanel.spec.plugin.kind,
+      spec: existingPanel.spec.plugin.spec, // TODO: Should we clone to be safe?
+    },
+    applyChanges: (values) => {
+      const { name, description, group, kind, spec } = values;
+      updatePanel(panelKey, {
+        kind: 'Panel',
+        spec: {
+          display: { name, description },
+          plugin: {
+            kind,
+            spec,
+          },
+        },
+      });
+
+      if (group !== groupIndex) {
+        // TO DO: need to move panel if panel group changes
+      }
+    },
+  };
+}
+
+/**
+ * Manages panel plugin spec state. The spec will be undefined while a plugin is being loaded.
+ */
+export function usePanelSpecState(panelPluginKind: string, initialState: unknown) {
+  // Keeping track of spec values by kind allows users to switch between panel types and come back with their old
+  // values intact from before the switch
+  const [specByKind, setSpecByKind] = useImmer<Record<string, unknown>>({ [panelPluginKind]: initialState });
+  const { data: plugin } = usePlugin('Panel', panelPluginKind, { enabled: panelPluginKind !== '' });
+  const pluginInitialSpec = useMemo(() => plugin?.createInitialOptions(), [plugin]);
+
+  // Use the value in specByKind or if unset, use the initial values from the plugin (which could still be undefined)
+  const spec = specByKind[panelPluginKind] ?? pluginInitialSpec;
+
+  // TODO: Do we want to expose more of a immer style API to plugin authors for managing their state, rather than the
+  // current "onChange" API?
```

Comment:
I'm definitely interested in hearing more specifics about this! I need to play around some more with the current API in TimeSeriesChartOptionsEditor, but this TODO sounds appealing.

Reply:
It seems like we only need to call setSpec/onSpecChange when user changes panel type though. If so, then the onChange api is more intuitive and simple. But if we have to call it every time some input changes, then i like the immer way!